bundles/org.openhab.binding.caddx/src/main/java/org/openhab/binding/caddx/internal/CaddxMessage.java

```diff
@@ -0,0 +1,504 @@
+    }
+
+    /**
+     * Constructor.
+     *
+     * @param message
+     *            - the message received
+     */
+    public CaddxMessage(byte[] message, boolean withChecksum) {
+        if (withChecksum && message.length < 3) {
+            throw new IllegalArgumentException("The message should be at least 3 bytes long");
+        }
+        if (!withChecksum && message.length < 1) {
+            throw new IllegalArgumentException("The message should be at least 1 byte long");
+        }
+
+        // Received data
+        byte[] msg = message;
+        if (withChecksum) {
+            checksum1In = message[message.length - 2];
+            checksum2In = message[message.length - 1];
+            msg = Arrays.copyOf(message, message.length - 2);
+        }
+
+        // Calculate the checksum
+        byte[] fletcherSum = fletcher(msg);
+        checksum1Calc = fletcherSum[0];
+        checksum2Calc = fletcherSum[1];
+        // Make the In checksum same as the Calculated in case it is not supplied
+        if (!withChecksum) {
+            checksum1In = checksum1Calc;
+            checksum2In = checksum2Calc;
+        }
+
+        this.message = msg;
+
+        // fill the message type
+        caddxMessageType = CaddxMessageType.valueOfMessageType((message[0] & 0x7f));
+
+        // Fill-in the properties
+        processCaddxMessage();
+    }
+
+    /**
+     * Builds a Caddx message for a zone bypass toggle command
+     *
+     * @param data The zone number
+     * @return The Caddx message object
+     */
+    public static CaddxMessage buildZoneBypassToggle(String data) {
+        byte[] arr = new byte[2];
+        arr[0] = 0x3f;
+        arr[1] = (byte) Integer.parseInt(data);
+
+        return new CaddxMessage(arr, false);
+    }
+
+    /**
+     * Builds a Caddx message for a zone status request command
+     *
+     * @param data The zone number
+     * @return The Caddx message object
+     */
+    public static CaddxMessage buildZoneStatusRequest(String data) {
+        byte[] arr = new byte[2];
+        arr[0] = 0x24;
+        arr[1] = (byte) Integer.parseInt(data);
+
+        return new CaddxMessage(arr, false);
+    }
+
+    /**
+     * Builds a Caddx message for a zone name request command
+     *
+     * @param data The zone number
+     * @return The Caddx message object
+     */
+    public static CaddxMessage buildZoneNameRequest(String data) {
+        byte[] arr = new byte[2];
+        arr[0] = 0x23;
+        arr[1] = (byte) Integer.parseInt(data);
+
+        return new CaddxMessage(arr, false);
+    }
+
+    /**
+     * Builds a Caddx message for a partition status request command
+     *
+     * @param data The partition number
+     * @return The Caddx message object
+     */
+    public static CaddxMessage buildPartitionStatusRequest(String data) {
+        byte[] arr = new byte[2];
+        arr[0] = 0x26;
+        arr[1] = (byte) Integer.parseInt(data);
+
+        return new CaddxMessage(arr, false);
```

Comment:
Why not refactor this into a new CaddxMessage constructor so it can be reused? `CaddxMessage(boolean, byte...) `would work no?

Reply:
I can refactor this into a new constructor CaddxMessage (CommandEnum, data) and put inside the whole logic of all the buildXxxxxxxx functions. I want the knowledge of the message building to be inside the CaddxMessage class. WDYT?

Reply:
That sounds fine, I'm more concerned about excessive code duplication.

Reply:
Class Refactored